base/compiler/tfuncs.jl

```diff
@@ -409,22 +409,41 @@ add_tfunc(pointerref, 3, 3,
           end, 4)
 add_tfunc(pointerset, 4, 4, (@nospecialize(a), @nospecialize(v), @nospecialize(i), @nospecialize(align)) -> a, 5)
 
+# more accurate typeof_tfunc for vararg tuples abstract only in length
+function typeof_concrete_vararg(@nospecialize(t))
+    np = length(t.parameters)
+    for i = 1:np
+        p = t.parameters[i]
+        if i == np && isvarargtype(p)
```

Comment:
I generally prefer a style that early exits in the error case, with the interesting case being as toplevel as possible, i.e.
```
function typeof_concrete_vararg(@nospecialize(t))
    np = length(t.parameters)
    for i = 1:np
        p = t.parameters[i]
        if i == np ? !isvarargtype(p) : !isconcretetype(p)
            return nothing
        end
     end
     pp = unwrap_unionall(t.parameters[end])
     if !isconcretetype(pp.parameters[1]) || !(pp.parameters[2] isa TypeVar)
         return nothing
     end
     return rewrap_unionall(Type{Tuple{t.parameters[1:np-1]..., pp}}, p)
end
```

up to you though.

Reply:
or sth like that. Not quite right, but you get the idea